src/amber/support/predicates.cr

```diff
@@ -0,0 +1,236 @@
+module Amber
+  module Support
+    module Predicates
+      module String
+        macro included
+            def str?
+                self.class == String
+            end
+
+            # email validation
+            def email?
+                Check.it self.match(/^[_]*([a-z0-9]+(\.|_*)?)+@([a-z][a-z0-9-]+(\.|-*\.))+[a-z]{2,6}$/)
+            end
+
+            # domain validation
+            def domain?
+                Check.it self.match(/^([a-z][a-z0-9-]+(\.|-*\.))+[a-z]{2,6}$/)
+            end
+
+            # url validation
+            def url?
+                Check.it self.match(/^(http(s)?(:\/\/))?(www\.)?[a-zA-Z0-9-_\.]+(\.[a-zA-Z0-9]{2,})([-a-zA-Z0-9:%_\+.~#?&\/\/=]*)/)
+            end
+
+            # ip v4 validation
+            def ipv4?
+                Check.it self.match(/^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/)
+            end
+
+            # ip v6 validation
+            def ipv6?
+                Check.it self.match(/^(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})$/)
+            end
+
+            # mac address validation
+            def mac_address?
+                Check.it self.match(/^([0-9a-fA-F][0-9a-fA-F]:){5}([0-9a-fA-F][0-9a-fA-F])$/)
+            end
+
+            # hex color validation
+            def hex_color?
+                Check.it self.match(/^#?([0-9A-F]{3}|[0-9A-F]{6})$/i)
+            end
+
+            # hexadecimal validation
+            def hex?
+                Check.it self.match(/^(0x)?[0-9A-F]+$/i)
+            end
+
+            # alpha characters validation
+            def alpha?(locale = "en-US")
+                matches = {
```

Comment:
This will allocate the hash each time the method is called => slow, you can put this outside the method (in a constant?) https://crystal-lang.org/docs/guides/performance.html#avoid-creating-temporary-objects-over-and-over